lib/IRGen/GenObjC.cpp

```diff
@@ -397,6 +397,42 @@ IRGenModule::getObjCProtocolGlobalVars(ProtocolDecl *proto) {
   return pair;
 }
 
+llvm::Constant *
+IRGenModule::getObjCProtocolRefSymRefDescriptor(ProtocolDecl *protocol) {
+  // See whether we already emitted this protocol reference.
+  auto found = ObjCProtocolSymRefs.find(protocol);
+  if (found != ObjCProtocolSymRefs.end()) {
+    return found->second;
+  }
+
+  ConstantInitBuilder InitBuilder(*this);
+  ConstantStructBuilder B(InitBuilder.beginStruct());
+  B.setPacked(true);
+  auto protocolRef = getAddrOfObjCProtocolRef(protocol, NotForDefinition);
+  B.addRelativeAddress(cast<llvm::Constant>(protocolRef.getAddress()));
+  auto ty = protocol->getDeclaredType()->getCanonicalType();
+  auto typeRef =
+      getTypeRef(ty, CanGenericSignature(), MangledTypeRefRole::FlatUnique)
+          .first;
+  B.addRelativeAddress(typeRef);
+  auto future = B.finishAndCreateFuture();
+
+  llvm::SmallString<64> nameBuffer;
+  StringRef protocolName = protocol->getObjCRuntimeName(nameBuffer);
+  auto *protocolSymRef = new llvm::GlobalVariable(
```

Comment:
Does this end up in `__objc_protorefs`? We'll want it there so the runtime can fix it up if this image doesn't end up with the canonical copy. It would also be good to have a test which uses a protocol from elsewhere, for example check that `NSObjectProtocol.self == objc_getProtocol("NSObject")`.

Reply:
`getAddrOfObjCProtocolRef(protocol, NotForDefinition)` does yes. `protocolSymRef` contains a relative reference to it.

Reply:
`getAddrOfObjCProtocolRef` is existing IRGen machinery to get a protocol reference that can be fixed up by the runtime

Reply:
Perfect.